Nil pointer panic in resolveTransportType for protocol scheme images

## Bug description

Running an MCP server via a protocol scheme (e.g., `npx://@zereight/mcp-gitlab`) causes a nil pointer panic in `resolveTransportType` at `cmd/thv/app/run_flags.go:463`.

The root cause is a nil-interface-vs-nil-pointer issue:

1. `pkg/runner/retriever/retriever.go` declares `var imageMetadata *types.ImageMetadata` (nil) in `handleProtocolScheme` and returns it without assignment (line 152)
2. `GetMCPServer` returns this nil `*ImageMetadata` as a `types.ServerMetadata` interface (line 127)
3. In `resolveTransportType`, the `serverMetadata != nil` check passes (the interface itself is non-nil), but `serverMetadata.GetTransport()` panics because the underlying pointer is nil

`resolveServerName` (line 473 in the same file) already handles this correctly using a type assertion with nil check:
```go
if imageMetadata, ok := serverMetadata.(*regtypes.ImageMetadata); ok && imageMetadata != nil {
```

## Steps to reproduce

```bash
thv run npx://@zereight/mcp-gitlab --name gitlab
```

Any `npx://`, `uvx://`, or `go://` protocol scheme that doesn't have a matching registry entry will trigger the panic.

## Expected behavior

The server should start normally, falling back to the default transport type when no metadata is available.

## Actual behavior

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x50 pc=0x102de6000]

goroutine 1 [running]:
github.com/stacklok/toolhive-core/registry/types.(*ImageMetadata).GetTransport(...)
    <autogenerated>:1
github.com/stacklok/toolhive/cmd/thv/app.resolveTransportType(...)
    cmd/thv/app/run_flags.go:463
```

## Suggested fix

Either apply the same type-assertion pattern that `resolveServerName` already uses:

```go
func resolveTransportType(runFlags *RunFlags, serverMetadata regtypes.ServerMetadata) string {
    if runFlags.Transport != "" {
        return runFlags.Transport
    }
    if imageMetadata, ok := serverMetadata.(*regtypes.ImageMetadata); ok && imageMetadata != nil {
        return imageMetadata.GetTransport()
    }
    return defaultTransportType
}
```

Or fix the source in `handleProtocolScheme` / `GetMCPServer` to return an explicit `nil` interface rather than a typed nil pointer when no metadata is available.

## Environment

- ToolHive version: latest main only — **this does not affect the v0.10.0 release**. The regression was introduced on main by #3912 (commit 80c3133f, merged 2026-02-20).
- OS: macOS (the code path is platform-independent)

## Additional context

This was introduced by #3912 (Move permissions and registry packages from toolhive to toolhive-core), which refactored the registry types. The `ServerMetadata` interface was introduced as part of that refactor, and the implicit nil-pointer-to-interface conversion in the return path wasn't caught.

This is a release blocker — any release cut from current main will panic on all protocol scheme images (`npx://`, `uvx://`, `go://`) that don't have a matching registry entry.

--- a/cmd/thv/app/run_flags.go
+++ b/cmd/thv/app/run_flags.go
@@ -439,7 +439,7 @@ func handleImageRetrieval(
 			return imageURL, serverMetadata, nil
 		}
 	}
-	return serverOrImage, nil, nil
+	return imageURL, nil, nil
 }
 
 // validateAndSetupProxyMode validates and sets default proxy mode if needed
@@ -454,13 +454,17 @@ func validateAndSetupProxyMode(runFlags *RunFlags) error {
 	return nil
 }
 
-// resolveTransportType selects the appropriate transport type based on flags and metadata
+// resolveTransportType selects the appropriate transport type based on flags and metadata.
+// Uses a type assertion with nil check to guard against typed nil pointers wrapped
+// in a non-nil interface (e.g., nil *ImageMetadata returned as ServerMetadata).
 func resolveTransportType(runFlags *RunFlags, serverMetadata regtypes.ServerMetadata) string {
 	if runFlags.Transport != "" {
 		return runFlags.Transport
 	}
-	if serverMetadata != nil {
-		return serverMetadata.GetTransport()
+	if imageMetadata, ok := serverMetadata.(*regtypes.ImageMetadata); ok && imageMetadata != nil {
+		if t := imageMetadata.GetTransport(); t != "" {
+			return t
+		}
 	}
 	return defaultTransportType
 }
@@ -547,7 +551,13 @@ func buildRunnerConfig(
 ) (*runner.RunConfig, error) {
 	transportType := resolveTransportType(runFlags, serverMetadata)
 	serverName := resolveServerName(runFlags, serverMetadata)
-	imageMetadata, _ := serverMetadata.(*regtypes.ImageMetadata)
+
+	// Use type assertion with nil check to guard against typed nil pointers
+	// wrapped in a non-nil interface (e.g., protocol scheme images).
+	var imageMetadata *regtypes.ImageMetadata
+	if md, ok := serverMetadata.(*regtypes.ImageMetadata); ok && md != nil {
+		imageMetadata = md
+	}
 
 	// Build default options
 	opts := []runner.RunConfigBuilderOption{
@@ -701,7 +711,7 @@ func configureMiddlewareAndOptions(
 func configureRemoteAuth(runFlags *RunFlags, serverMetadata regtypes.ServerMetadata) ([]runner.RunConfigBuilderOption, error) {
 	var opts []runner.RunConfigBuilderOption
 
-	if remoteServerMetadata, ok := serverMetadata.(*regtypes.RemoteServerMetadata); ok {
+	if remoteServerMetadata, ok := serverMetadata.(*regtypes.RemoteServerMetadata); ok && remoteServerMetadata != nil {
 		remoteAuthConfig, err := getRemoteAuthFromRemoteServerMetadata(remoteServerMetadata, runFlags)
 		if err != nil {
 			return nil, err

--- a/cmd/thv/app/run_flags_test.go
+++ b/cmd/thv/app/run_flags_test.go
@@ -15,6 +15,7 @@ import (
 	"github.com/stretchr/testify/require"
 
 	"github.com/stacklok/toolhive-core/logging"
+	regtypes "github.com/stacklok/toolhive-core/registry/types"
 	"github.com/stacklok/toolhive/pkg/config"
 )
 
@@ -401,4 +402,110 @@ func TestBuildRunnerConfig_TelemetryProcessing_Integration(t *testing.T) {
 	assert.Equal(t, false, finalInsecure, "Insecure setting should use runFlags value when not set via CLI")
 	assert.Equal(t, true, finalUseLegacyAttributes, "UseLegacyAttributes should default to true when not set via CLI or config")
 	assert.Equal(t, false, finalEnablePrometheusMetricsPath, "Enable Prometheus metrics path should use runFlags value when not set via CLI")
+}
+
+func TestResolveTransportType(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name           string
+		runFlags       *RunFlags
+		serverMetadata regtypes.ServerMetadata
+		expected       string
+	}{
+		{
+			name:           "explicit transport flag takes precedence",
+			runFlags:       &RunFlags{Transport: "stdio"},
+			serverMetadata: &regtypes.ImageMetadata{BaseServerMetadata: regtypes.BaseServerMetadata{Transport: "sse"}},
+			expected:       "stdio",
+		},
+		{
+			name:           "transport from metadata when flag is empty",
+			runFlags:       &RunFlags{},
+			serverMetadata: &regtypes.ImageMetadata{BaseServerMetadata: regtypes.BaseServerMetadata{Transport: "sse"}},
+			expected:       "sse",
+		},
+		{
+			name:           "nil interface returns default transport",
+			runFlags:       &RunFlags{},
+			serverMetadata: nil,
+			expected:       defaultTransportType,
+		},
+		{
+			name:           "typed nil pointer in interface returns default (protocol scheme case)",
+			runFlags:       &RunFlags{},
+			serverMetadata: regtypes.ServerMetadata((*regtypes.ImageMetadata)(nil)),
+			expected:       defaultTransportType,
+		},
+		{
+			name:           "metadata with empty transport returns default",
+			runFlags:       &RunFlags{},
+			serverMetadata: &regtypes.ImageMetadata{},
+			expected:       defaultTransportType,
+		},
+		{
+			name:           "explicit flag overrides even with nil metadata",
+			runFlags:       &RunFlags{Transport: "streamable-http"},
+			serverMetadata: nil,
+			expected:       "streamable-http",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			result := resolveTransportType(tt.runFlags, tt.serverMetadata)
+			assert.Equal(t, tt.expected, result)
+		})
+	}
+}
+
+func TestResolveServerName(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name           string
+		runFlags       *RunFlags
+		serverMetadata regtypes.ServerMetadata
+		expected       string
+	}{
+		{
+			name:           "explicit name flag takes precedence",
+			runFlags:       &RunFlags{Name: "my-server"},
+			serverMetadata: &regtypes.ImageMetadata{BaseServerMetadata: regtypes.BaseServerMetadata{Name: "registry-name"}},
+			expected:       "my-server",
+		},
+		{
+			name:           "name from metadata when flag is empty",
+			runFlags:       &RunFlags{},
+			serverMetadata: &regtypes.ImageMetadata{BaseServerMetadata: regtypes.BaseServerMetadata{Name: "registry-name"}},
+			expected:       "registry-name",
+		},
+		{
+			name:           "nil interface returns empty string",
+			runFlags:       &RunFlags{},
+			serverMetadata: nil,
+			expected:       "",
+		},
+		{
+			name:           "typed nil pointer in interface returns empty string (protocol scheme case)",
+			runFlags:       &RunFlags{},
+			serverMetadata: regtypes.ServerMetadata((*regtypes.ImageMetadata)(nil)),
+			expected:       "",
+		},
+		{
+			name:           "explicit flag overrides even with nil metadata",
+			runFlags:       &RunFlags{Name: "explicit"},
+			serverMetadata: nil,
+			expected:       "explicit",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			result := resolveServerName(tt.runFlags, tt.serverMetadata)
+			assert.Equal(t, tt.expected, result)
+		})
+	}
 }

--- a/pkg/api/v1/workload_service.go
+++ b/pkg/api/v1/workload_service.go
@@ -192,7 +192,7 @@ func (s *WorkloadService) BuildFullRunConfig(
 			return nil, fmt.Errorf("failed to retrieve MCP server image: %w", err)
 		}
 
-		if remoteServerMetadata, ok := serverMetadata.(*regtypes.RemoteServerMetadata); ok {
+		if remoteServerMetadata, ok := serverMetadata.(*regtypes.RemoteServerMetadata); ok && remoteServerMetadata != nil {
 			if remoteServerMetadata.OAuthConfig != nil {
 				// Default resource: user-provided > registry metadata > derived from remote URL
 				resource := req.OAuthConfig.Resource
@@ -228,8 +228,11 @@ func (s *WorkloadService) BuildFullRunConfig(
 				}
 			}
 		}
-		// Handle server metadata - API only supports container servers
-		imageMetadata, _ = serverMetadata.(*regtypes.ImageMetadata)
+		// Handle server metadata - API only supports container servers.
+		// Use type assertion with nil check to guard against typed nil pointers.
+		if md, ok := serverMetadata.(*regtypes.ImageMetadata); ok && md != nil {
+			imageMetadata = md
+		}
 	}
 
 	// Build RunConfig
@@ -291,8 +294,10 @@ func (s *WorkloadService) BuildFullRunConfig(
 	transportType := "streamable-http"
 	if req.Transport != "" {
 		transportType = req.Transport
-	} else if serverMetadata != nil {
-		transportType = serverMetadata.GetTransport()
+	} else if md, ok := serverMetadata.(*regtypes.ImageMetadata); ok && md != nil {
+		if t := md.GetTransport(); t != "" {
+			transportType = t
+		}
 	}
 
 	// Configure middleware from flags

--- a/pkg/mcp/server/run_server.go
+++ b/pkg/mcp/server/run_server.go
@@ -50,8 +50,12 @@ func (h *Handler) RunServer(ctx context.Context, request mcp.CallToolRequest) (*
 		return mcp.NewToolResultError(fmt.Sprintf("Failed to get MCP server: %v", err)), nil
 	}
 
-	// Build run configuration
-	imageMetadata, _ := serverMetadata.(*types.ImageMetadata)
+	// Build run configuration.
+	// Use type assertion with nil check to guard against typed nil pointers.
+	var imageMetadata *types.ImageMetadata
+	if md, ok := serverMetadata.(*types.ImageMetadata); ok && md != nil {
+		imageMetadata = md
+	}
 
 	runConfig, err := buildServerConfig(ctx, args, imageURL, imageMetadata)
 	if err != nil {

--- a/pkg/runner/retriever/retriever.go
+++ b/pkg/runner/retriever/retriever.go
@@ -73,7 +73,7 @@ func GetMCPServer(
 		slog.Debug("Attempting to retrieve MCP server from protocol scheme",
 			"server_or_image", serverOrImage)
 		var err error
-		imageToUse, imageMetadata, err = handleProtocolScheme(ctx, serverOrImage, rawCACertPath, imageManager, runtimeOverride)
+		imageToUse, err = handleProtocolScheme(ctx, serverOrImage, rawCACertPath, imageManager, runtimeOverride)
 		if err != nil {
 			return "", nil, err
 		}
@@ -124,32 +124,34 @@ func GetMCPServer(
 		return "", nil, fmt.Errorf("failed to retrieve or pull image: %w", err)
 	}
 
-	return imageToUse, imageMetadata, nil
+	// Guard against returning a typed nil pointer as a ServerMetadata interface.
+	// A nil *ImageMetadata wrapped in a non-nil interface would cause callers
+	// checking "serverMetadata != nil" to proceed and panic on method calls.
+	if imageMetadata != nil {
+		return imageToUse, imageMetadata, nil
+	}
+	return imageToUse, nil, nil
 }
 
-// handleProtocolScheme handles the protocol scheme case
+// handleProtocolScheme handles the protocol scheme case.
+// Protocol schemes (npx://, uvx://, go://) don't have registry metadata,
+// so this only returns the generated image name.
 func handleProtocolScheme(
 	ctx context.Context,
 	serverOrImage string,
 	rawCACertPath string,
 	imageManager images.ImageManager,
 	runtimeOverride *templates.RuntimeConfig,
-) (string, *types.ImageMetadata, error) {
-	var imageMetadata *types.ImageMetadata
-	var imageToUse string
-
+) (string, error) {
 	slog.Debug("Detected protocol scheme", "server", serverOrImage)
 	// Process the protocol scheme and build the image
 	caCertPath := resolveCACertPath(rawCACertPath)
 	generatedImage, err := runner.HandleProtocolScheme(ctx, imageManager, serverOrImage, caCertPath, runtimeOverride)
 	if err != nil {
-		return "", nil, errors.Join(ErrBadProtocolScheme, err)
+		return "", errors.Join(ErrBadProtocolScheme, err)
 	}
-	// Update the image in the runConfig with the generated image
 	slog.Debug("Using built image", "image", generatedImage, "original", serverOrImage)
-	imageToUse = generatedImage
-
-	return imageToUse, imageMetadata, nil
+	return generatedImage, nil
 }
 
 // handleGroupLookup handles the group lookup case

